### PR TITLE
Chan

### PR DESCRIPTION
**Pull request by chanpham97**
send playlist info to db and links to unique playlist url

### CODE REVIEW

File: script/db/index.js [hunk 1]
```diff
@@ -61,6 +60,28 @@ module.exports = firebase => {
...
+        return snapshot;
+      });
+    },
+
+    /**
+     * @param  string uid of the user being sought after
+     * @return dictionary the user data in the form of the tree structure
+     */
+    readUserData: (uid) => {
+
```
Comment (alanhdu) [inline script/db/index.js]: Unnecessary new line.

File: script/db/index.js [hunk 2]
```diff
@@ -61,6 +60,28 @@ module.exports = firebase => {
     },
 
     /**
+     * @param string playlistID playlistID of the playlist being sought after
```
Comment (alanhdu) [inline script/db/index.js]: `playlistID playlistID of the playlist`? Sounds a little repetitive.

File: script/db/index.js [hunk 3]
```diff
@@ -61,6 +60,28 @@ module.exports = firebase => {
...
+     * @return dictionary the playlist data in the form of the tree structure
```
Comment (alanhdu) [inline script/db/index.js]: What does "tree structure" mean here?

File: script/db/index.js [hunk 4]
```diff
@@ -151,6 +172,7 @@ module.exports = firebase => {
                 usersRef.child(uid).update({
                   playListIDs: newList
                 });
+                return playListKey;
               });
               return;
```
Comment (alanhdu) [inline script/db/index.js]: Seems like a spurious extra return?

File: script/db/index.js [hunk 5]
```diff
@@ -101,28 +122,28 @@ module.exports = firebase => {
...
+          return playListRef.push(playListInput).key;;
         }
 
-        usersRef.child(uid).child("playListIDs").once("value", snapshot => {
+        return usersRef.child(uid).child("playListIDs").once("value").then(snapshot => {
           const userPlaylists = snapshot.val();
 
-          playListRef.once("value", allPlaylists => {
+          return playListRef.once("value").then(allPlaylists => {
             const createPlayList = (() => {
```
Comment (alanhdu) [inline script/db/index.js]: Why are we declaring an anonymous function just to call it immediately? Just execute the code inline.

Also, maybe rename to `shouldCreatePlaylist` ... might be clearer.

File: script/db/index.js [hunk 6]
```diff
@@ -142,7 +163,7 @@ module.exports = firebase => {
             })();
```
Comment (alanhdu) [inline script/db/index.js]: This function looks over-complicated. Does:

```javascript
return userPlaylists.map(x => playListData[x].urls).every(currURLs => {
  if (currURLS.length !== urls.length) {
    return true
  }
  return !currURLS.every((cu, i) => cu == urls[i])
});
````

work?

File: script/db/index.js [hunk 7]
```diff
@@ -151,6 +172,7 @@ module.exports = firebase => {
                 usersRef.child(uid).update({
                   playListIDs: newList
```
Comment (alanhdu) [inline script/db/index.js]: Does this need special casing in case `newList` is empty and we should be inserting our sentinel value instead?

(TBH, I'm still not a fan of this sentinel value business. I honestly think it'd be easier to store `null` or just not have a key instead of using `["filter"]`).

File: script/db/index.js [hunk 8]
```diff
@@ -101,28 +122,28 @@ module.exports = firebase => {
...
+      return usersRef.child(uid).once("value").then(snapshot => {
+        if (uid === "guest" && userName === "guest") {
           console.error(`ERROR: A playList was created under the user ${userName} but the user does not exist`);
-
-          return;
+          return playListRef.push(playListInput).key;;
         }
 
-        usersRef.child(uid).child("playListIDs").once("value", snapshot => {
+        return usersRef.child(uid).child("playListIDs").once("value").then(snapshot => {
```
Comment (alanhdu) [inline script/db/index.js]: I'm confused what's going on here. Why are we nesting promises? In fact, why does the outer promise (`usersRef.child(uid).once("value")`) exist at all when we don't seem to be referencing it's snapshot?

File: script/db/index.js [hunk 9]
```diff
@@ -142,7 +163,7 @@ module.exports = firebase => {
             })();
 
             if (createPlayList) {
-              usersRef.child(uid).child("playListIDs").once("value").then(snapshot => {
+              return usersRef.child(uid).child("playListIDs").once("value").then(snapshot => {
```
Comment (alanhdu) [inline script/db/index.js]: Don't we already have this value (see L141)? Why do we need another promise?

File: script/db/index.js [hunk 10]
```diff
@@ -101,28 +122,28 @@ module.exports = firebase => {
...
       const playListInput = {
         userName: userName,
         title: title,
         description: description,
         urls: urls
       };
 
-      usersRef.child(uid).once("value", snapshot => {
-        if (snapshot.val() === null) {
+      return usersRef.child(uid).once("value").then(snapshot => {
```
Comment (alanhdu) [inline script/db/index.js]: It's super hard to follow the logic here (not your fault -- it's one of the well-known weakpoints of futures, which is why everyone seems to be converging to .NET style `async`/`await`).

Still, is there a way to unnest some of the promise chaining? I'm not super familiar with javascript promises, but something like:

```javascript
if (uid === "guest" && userName === "guest") {
  console.error(`ERROR: A playList was created under the user ${userName} but the user does not exist`);
  return playListRef.push(playListInput).key;
}
const uPlaylistPromise = usersRef.child(uid).child("playListIDs")
  .once("value")
  .then(snapshot => snapshot.val());

return Promise.all([uPlaylistPromise, playListRef.once("value")])
  .then([userPlaylists, allPlaylists] => {
    let shouldCreatePlayList = false;
    if (userPlaylists.length === 0 || userPlaylists[0] === EMPTY) {
      return true; // Why does this need special casing?
    }   
    const playListData = allPlaylists.val()
    shouldCreatePlayList = userPlaylists
      .map(x => playListData[x].urls)
      .every(currURLs => currURLs !== urls); // Does this work here?

    if (!shouldCreatePlayList) {
      console.error(`ERROR: The user ${userName} already has this exact playList`);
      return;
    } 

    const newList = userPlaylists[0] === EMPTY ? [] : userPlaylists.slice();
    const playListKey = playListRef.push(playListInput).key;

    newList.push(playListKey);
    usersRef.child(uid).update({
      playListIDs: newList
    }); 
    return playListKey;
  }); 
```

by using `Promise.all` and refactoring a little, we reduce how many nested callbacks we need which IMO make sit much easier to read.

File: script/db/index.js [hunk 11]
```diff
@@ -142,7 +163,7 @@ module.exports = firebase => {
             })();
 
             if (createPlayList) {
-              usersRef.child(uid).child("playListIDs").once("value").then(snapshot => {
+              return usersRef.child(uid).child("playListIDs").once("value").then(snapshot => {
                 const tempList = snapshot.val();
                 const newList = tempList[0] === EMPTY ? [] : tempList.slice();
                 const playListKey = playListRef.push(playListInput).key;
```
Comment (alanhdu) [inline script/db/index.js]: Does this need special casing in case `playListInput` is empty? 

(TBH, I'm still not a fan of this sentinel value business. I honestly think it'd be easier to store `null` or just not have a key instead of using `["filter"]`).

File: script/routes/index.js [hunk 12]
```diff
@@ -12,6 +12,31 @@ module.exports = (app, db) => {
...
+  app.get("/make", (req, res) => {
+    res.render("make");
+  });
+
+  app.post("/create", (req, res) => {
+    const {
+      title,
+      description,
+      fields
+    } = req.body;
```
Comment (alanhdu) [inline script/routes/index.js]: Can this be one line?

File: script/routes/index.js [hunk 13]
```diff
@@ -12,6 +12,31 @@ module.exports = (app, db) => {
...
+    console.log(req.body);
```
Comment (alanhdu) [inline script/routes/index.js]: Extraneous logging statement

File: script/routes/index.js [hunk 14]
```diff
@@ -12,6 +12,31 @@ module.exports = (app, db) => {
...
+
+    const user = req.session.user;
+    const [uid, username] = user === undefined ? ["guest", "guest"] : [user.uid, user.username];
```
Comment (alanhdu) [inline script/routes/index.js]: Wait... what's the difference between `uid` and `user.username`? Isn't `uid` an integer or something?

File: template/create.jade [hunk 15]
```diff
@@ -18,11 +19,12 @@ body
...
\ No newline at end of file
+				label Links:
+				div(class="links")
+					div(class="entry input-group")
+						input(class="form-control", name="fields", type="text", placeholder="Add URL")
+						span(class="input-group-btn")
+							button(class="btn btn-success btn-add", type="button")
+								span(class="glyphicon glyphicon-plus")
+			br
+			input(class="btn btn-default", id="createButton", type="submit" value="Create Tape")
```
Comment (alanhdu) [inline template/create.jade]: Could we be consistent about tabs vs spaces? The other jade file uses two space indent, while this uses tabs.

File: script/db/index.js (hunk 6 above)
Comment (alanhdu) [inline script/db/index.js]: Also, do we need to sort the URLs before comparing? Are they already sorted? Do we need to normalize them somehow (e.g. two different urls might redirect to the same page)?

It's unclear to me that just an array of URLs really captures all the invariants we probably want.

File: script/routes/index.js (hunk 14 above)
Comment (chanpham97) [inline script/routes/index.js]: yeah user id is the bash value

File: template/create.jade (hunk 15 above)
Comment (chanpham97) [inline template/create.jade]: will fix!

File: script/routes/index.js [hunk 16]
```diff
@@ -4,8 +4,37 @@ const firebase = require("firebase");
...
-  app.get("/", (_, res) => {
-    res.render("index");
+  app.get("/", (req, res) => {
+    if (req.session && req.session.user) {
+      console.log("in session");
+      const user = req.session.user;
+
+      db.readUserData(user.uid)
+        .then(userData => {
+          //TODO: don't include the password @Nick will change this
```
Comment (alanhdu) [inline script/routes/index.js]: Spurious `TODO`?

File: script/routes/index.js [hunk 17]
```diff
@@ -4,8 +4,37 @@ const firebase = require("firebase");
...
+          const {
+            userName
+          } = userData.val();
```
Comment (alanhdu) [inline script/routes/index.js]: One line please

File: script/routes/index.js [hunk 18]
```diff
@@ -4,8 +4,37 @@ const firebase = require("firebase");
...
+
+    // if (user) {
+      
+    //   res.render("index", { 
+    //     userName: req.session.user.username
+    //   });
+    // } else {
+    //   console.log("not in session");
+    //   res.render("index");
+    // }
```
Comment (alanhdu) [inline script/routes/index.js]: Delete extraneous comments

File: template/user.jade [hunk 19]
```diff
@@ -7,15 +7,15 @@ form(action="/logout", method="post")
...
+//-     const socket = io('http://localhost:3000');
+//-     socket.on("hello", function (data) {
+//-       //console.log(data);
+//-       data.hello.forEach(function (el) {
+//-         socket.emit("getPlayList", {pl: el});
+//-       })
+//-     });
+//-     socket.on("playlistinfo", function (data) {
+//-       console.log(data);
+//-     })
```
Comment (alanhdu) [inline template/user.jade]: Delete extraneous comments.

File: public/js/create.js [hunk 20]
```diff
@@ -7,7 +7,7 @@ $(function() {
     e.preventDefault();
     var controlForm = $(".controls form:first");
     var currentEntry = $(this).parents(".entry:first");
-    var newEntry = $(currentEntry.clone()).appendTo(controlForm);
+    var newEntry = $(currentEntry.clone()).appendTo(".links");
```
Comment (Howon) [inline public/js/create.js]: ?
